src/components/structures/RoomView.js

```diff
@@ -180,13 +152,28 @@ module.exports = React.createClass({
         if (this.unmounted) {
             return;
         }
-        this.setState({
+        const newState = {
             roomId: RoomViewStore.getRoomId(),
             roomAlias: RoomViewStore.getRoomAlias(),
             roomLoading: RoomViewStore.isRoomLoading(),
             roomLoadError: RoomViewStore.getRoomLoadError(),
             joining: RoomViewStore.isJoining(),
-        }, () => {
+            eventId: RoomViewStore.getEventId(),
+            eventPixelOffset: RoomViewStore.getEventPixelOffset(),
+            isEventHighlighted: RoomViewStore.isEventHighlighted(),
+        };
+
+        if (this.state.eventId !== newState.eventId) {
+            newState.searchResults = null;
+        }
+
+        // Store the scroll state for the previous room so that we can return to this
+        // position when viewing this room in future.
+        if (this.state.roomId !== newState.roomId) {
+            this._updateScrollMap(this.state.roomId);
+        }
+
+        this.setState(newState, () => {
             this._onHaveRoom();
```

Comment:
So basically yes, but I shall leave a comment explaining so.
 - `onRoom ` will nop
 - `_onHaveRoom` will possibly peek but only if the user is not already joining.